client/dashboard/src/pages/toolsets/ResourcesTab.tsx

```diff
@@ -0,0 +1,255 @@
+      request: {
+        slug: toolset.slug,
+        updateToolsetRequestBody: {
+          resourceUrns: [...currentResourceUrns, resourceUrn],
+        },
+      },
+    });
+  };
+
+  const removeResourceFromToolset = (resourceUrn: string) => {
+    updateToolsetMutation.mutate({
+      request: {
+        slug: toolset.slug,
+        updateToolsetRequestBody: {
+          resourceUrns: currentResourceUrns.filter(
+            (urn) => urn !== resourceUrn,
+          ),
+        },
+      },
+    });
+  };
+
+  // Show empty state if no resources exist at all in the system
+  if (allResources.length === 0) {
+    return (
+      <div className="flex items-center justify-center py-12">
+        <Type muted>No current sources provide resources</Type>
+      </div>
+    );
+  }
+
+  return (
+    <Cards>
+      {toolsetResources
+        .filter((resource) => resource.type === "function")
+        .map((resource) => {
+          const functionName = functionIdToName?.[resource.functionId];
+          return (
+            <ResourceCard
+              key={resource.resourceUrn}
+              resource={resource}
+              functionName={functionName}
+              onDelete={() => removeResourceFromToolset(resource.resourceUrn)}
+            />
+          );
+        })}
+      {allResources && allResources.length > 0 && (
+        <ResourceSelectPopover
+          allResources={allResources}
+          currentResourceUrns={currentResourceUrns}
+          onSelect={(resourceUrn) => addResourceToToolset(resourceUrn)}
+        />
+      )}
+    </Cards>
+  );
+}
+
+function ResourceCard({
+  resource,
+  functionName,
+  onDelete,
+}: {
+  resource: Resource;
+  functionName?: string;
+  onDelete: () => void;
+}) {
+  const actions = [
+    {
+      label: "Remove from toolset",
+      onClick: onDelete,
+      icon: "trash" as const,
+      destructive: true,
+    },
+  ];
+
+  return (
+    <div className="border border-border rounded-lg p-4 flex items-start gap-4">
+      <div className="p-2 rounded-md bg-muted shrink-0">
+        <Newspaper className="size-5 text-muted-foreground" strokeWidth={1.5} />
+      </div>
+      <div className="flex-1 min-w-0">
+        <Stack gap={1}>
+          <Stack direction="horizontal" justify="space-between" align="start">
+            <div className="text-sm font-medium truncate">
+              {resource.name}
+              {functionName && (
+                <span className="text-xs text-muted-foreground font-normal ml-1">
+                  ({functionName})
+                </span>
+              )}
+            </div>
+            <MoreActions actions={actions} />
+          </Stack>
+          <Type small muted className="font-mono truncate">
+            {resource.uri}
+          </Type>
+        </Stack>
+      </div>
+    </div>
+  );
+}
+
+function ResourceSelectPopover({
+  allResources,
+  currentResourceUrns,
+  onSelect,
+}: {
+  allResources: GeneratedResource[];
```

Comment:
Let's try to avoid using this type? Like we should be writing this anywhere in the frontend `r.functionResourceDefinition?.resourceUrn`

Reply:
I can add a unionized version of this